File: standalone/dzen2-hlwm/lua/pipehandler.04-event.lua
-- luaposix available in AUR
local posix = require "posix"

local common = require('.common')
local helper = require('.helper')
local output = require('.output')

local _M = {}

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- helper

-- because os.clock function will hogs your cpu
function _M.os_sleep(n)
  os.execute('sleep ' .. tonumber(n))
end

-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- pipe

function _M.handle_command_event(monitor, event)
    -- find out event origin
    column = common.split(event, "\t")
    origin = column[1] -- non zero based

    tag_cmds = {'tag_changed', 'tag_flags', 'tag_added', 'tag_removed'}
    title_cmds = {'window_title_changed', 'focus_changed'}

    if origin == 'reload' then
        os.execute('pkill dzen2')
    elseif origin == 'quit_panel' then
        os.exit()
    elseif common.has_value(tag_cmds, origin) then
        output.set_tag_value(monitor)
    elseif common.has_value(title_cmds, origin) then
        output.set_windowtitle(column[3])
    elseif origin == 'interval' then
        output.set_datetime()
    end
end

function _M.content_init(monitor, pipe_dzen2_out)
    -- initialize statusbar before loop
    output.set_tag_value(monitor)
    output.set_windowtitle('')
    output.set_datetime()

    local text = output.get_statusbar_text(monitor)
    pipe_dzen2_out:write(text .. "\n")
    pipe_dzen2_out:flush()
end

function _M.content_event_idle(pipe_cat_out)
    local pid = posix.fork()

    if pid == 0 then -- this is the child process
        -- start a pipe
        command_in = 'herbstclient --idle'
        local pipe_in  = assert(io.popen(command_in,  'r'))
  
        -- wait for each event 
        for event in pipe_in:lines() do
            posix.write(pipe_cat_out, event)
            io.flush()
        end -- for loop
   
        pipein:close()
    else             -- this is the parent process
        -- nothing
    end
end

function _M.content_event_interval(pipe_cat_out) 
    local pid = posix.fork()

    if pid == 0 then -- this is the child process
        while true do
            posix.write(pipe_cat_out, "interval\n")
            io.flush() 

            _M.os_sleep(1)
        end
    else             -- this is the parent process
        -- nothing
    end
end

function _M.content_walk(monitor, pipe_dzen2_out)  
    rd, wr = posix.pipe()

    _M.content_event_idle(wr)
    _M.content_event_interval(wr)

    local bufsize = 4096
    local event = ''

    while true do
        -- wait for next event, trim newline
        event = common.trim1(posix.read(rd, bufsize))
        if event == nil or #event == 0 then break end
    
        _M.handle_command_event(monitor, event)    
    
        text = output.get_statusbar_text(monitor)
        pipe_dzen2_out:write(text .. "\n")
        pipe_dzen2_out:flush()
    end -- not using for loop

    posix.close(rd)
    posix.close(wr)
end

function _M.run_dzen2(monitor, parameters) 
    local command_out    = 'dzen2 ' .. parameters
    local pipe_dzen2_out = assert(io.popen(command_out, 'w'))
    
    _M.content_init(monitor, pipe_dzen2_out)
    _M.content_walk(monitor, pipe_dzen2_out) -- loop for each event
        
    pipe_dzen2_out:close()
end

function _M.detach_dzen2(monitor, parameters)
    local pid_dzen2 = posix.fork()

    if pid_dzen2 == 0 then -- this is the child process
        _M.run_dzen2(monitor, parameters)
    else             -- this is the parent process
        -- nothing
    end
end

function _M.detach_transset()
    local pid_transset = posix.fork()

    if pid_transset == 0 then -- this is the child process
        common.sleep(1)
        os.execute('transset .8 -n dzentop >/dev/null') 
    else                      -- this is the parent process
        -- nothing
    end
end


-- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- -----
-- return

return _M

File: standalone/dzen2-hlwm/ruby/pipehandler.04-event.rb
require_relative 'output'

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# pipe

def handle_command_event(monitor, event) 
  # find out event origin
  column = event.split("\t")
  origin = column[0]
    
  tag_cmds = ['tag_changed', 'tag_flags', 'tag_added', 'tag_removed']
  title_cmds = ['window_title_changed', 'focus_changed']

  case origin
  when 'reload'
    os.system('pkill dzen2')
  when 'quit_panel'
    exit
  when *tag_cmds       # splat operator
    set_tag_value(monitor)
  when *title_cmds     # splat operator
    title = column.length > 2 ? column[2] : ''
    set_windowtitle(title)
  when 'interval'
    set_datetime()
  end
end

def content_init(monitor, dzen2_stdin)
  # initialize statusbar before loop
  set_tag_value(monitor)
  set_windowtitle('')
  set_datetime()
      
  text = get_statusbar_text(monitor)
  dzen2_stdin.puts(text)
end

def content_event_idle(cat_stdin)
  pid = fork do 
    # start an io
    command_in = 'herbstclient --idle'
  
    IO.popen(command_in, "r") do |io_idle|
      while io_idle do 
         # read next event
        event = io_idle.gets
        cat_stdin.print(event)
      end
      io_idle.close()
    end
  end

  Process.detach(pid)  
end


def content_event_interval(cat_stdin)
  pid = fork do 
    while true do
      cat_stdin.print "interval\n"

      sleep(1)
    end
  end

  Process.detach(pid)  
end

def content_walk(monitor, dzen2_stdin)
  # note the r+ mode for bidirectional
  IO.popen('cat', 'r+') do |io_cat| 

    content_event_idle(io_cat)
    content_event_interval(io_cat)

    while io_cat do 
      # read next event, trim newline
      event = (io_cat.gets).strip
      handle_command_event(monitor, event)
        
      text = get_statusbar_text(monitor)
      dzen2_stdin.puts(text)
    end
  
  io_cat.close()
  end
end

def run_dzen2(monitor, parameters)
  command_out  = 'dzen2 ' + parameters
  IO.popen(command_out, "w") do |io_dzen2| 
    content_init(monitor, io_dzen2)
    content_walk(monitor, io_dzen2) # loop for each event
        
    io_dzen2.close()    
  end
end

def detach_dzen2(monitor, parameters)
  # warning: Signal.trap is application wide
  Signal.trap("PIPE", "EXIT")
    
  pid_dzen2 = fork { run_dzen2(monitor, parameters) }
  Process.detach(pid_dzen2)
end

def detach_transset()
  pid_transset = fork do
    sleep(1)
    system('transset .8 -n dzentop >/dev/null')        
  end
    
  Process.detach(pid_transset)
end
